File: TypeDefinitions/PlayerProxy.cs
namespace WorldServer.Scripting.MoonSharp.Proxies;

using Common.Enums.GameData;
using Common.Enums.SystemData;
using global::MoonSharp.Interpreter;
using JetBrains.Annotations;
using World.Objects;

[PublicAPI]
public class PlayerProxy : UnitProxy
{
    private readonly Player _player;

    [MoonSharpHidden]
    public PlayerProxy(Player player)
        : base(player)
    {
        _player = player;
    }

    #region Money

    [PublicAPI]
    public uint Money => _player.GetMoney();

    [PublicAPI]
    public void AddMoney(uint money) => _player.AddMoney(money);

    [PublicAPI]
    public bool RemoveMoney(uint money) => _player.RemoveMoney(money);

    #endregion

    #region Items

    [PublicAPI]
    public bool AddItem(uint itemId, ushort count) => _player.Items.CreateItemNotify(itemId, count) == ItemResult.ResultOk;

    [PublicAPI]
    public bool RemoveItem(uint itemId, ushort count) => _player.Items.RemoveItems(itemId, count);

    [PublicAPI]
    public bool HasItem(uint itemId, ushort count) => _player.Items.HasItemCountInInventory(itemId, count);

    #endregion

    #region Chat

    [PublicAPI]
    public void SendLocalizeString(string message,  ChatLogFilters filter, LocalizedText localizeEntry) => _player.SendLocalizeString(message, (ChatLogFilter)filter, localizeEntry);

    #endregion

    #region World

    [PublicAPI]
    public void Teleport(ushort zoneId, uint x, uint y, ushort z, ushort o) => _player.Teleport(zoneId, x, y, z, o);

    #endregion

    #region Quest

    [PublicAPI]
    public bool TriggerQuestObjective(ushort questId, ushort questObjectiveId) => _player.Quests.TriggerQuestObjective(questId, questObjectiveId);

    #endregion

    #region ActionCounters

    [PublicAPI]
    public void IncrementActionCounter(ushort actionCounterId, uint value) => _player.ActionCounters.Increment(actionCounterId, value);

    #endregion
}

File: TypeDefinitions/PlayerProxy.lua
-- player_proxy.lua

---@class PlayerProxy : UnitProxy
---@field Money number
local PlayerProxy = {}

--- Adds money to the player.
---@param money number
function PlayerProxy:AddMoney(money) end

--- Removes money from the player.
---@param money number
---@return boolean
function PlayerProxy:RemoveMoney(money) end

--- Adds an item to the player's inventory.
---@param itemId number
---@param count number
---@return boolean
function PlayerProxy:AddItem(itemId, count) end

--- Removes an item from the player's inventory.
---@param itemId number
---@param count number
---@return boolean
function PlayerProxy:RemoveItem(itemId, count) end

--- Checks if the player has a certain amount of an item in their inventory.
---@param itemId number
---@param count number
---@return boolean
function PlayerProxy:HasItem(itemId, count) end

--- Sends a localized string to the player's chat.
---@param message string
---@param filter ChatLogFilters
---@param localizeEntry LocalizedText
function PlayerProxy:SendLocalizeString(message, filter, localizeEntry) end

--- Teleports the player to a specified location.
---@param zoneId number
---@param x number
---@param y number
---@param z number
---@param o number
function PlayerProxy:Teleport(zoneId, x, y, z, o) end

--- Triggers a quest objective for the player.
---@param questId number
---@param questObjectiveId number
---@return boolean
function PlayerProxy:TriggerQuestObjective(questId, questObjectiveId) end

return PlayerProxy
